server/db/db.rs

```diff
@@ -1523,6 +1523,37 @@ impl LockedDatabase {
         }
     }
 
+    /// Transfer given recordings to garbage and refresh in-memory map
+    /// of files to garbage collect.
+    pub fn delete_recordings(
+        &mut self,
+        stream_id: i32,
+        recording_ids: Range<i32>,
+    ) -> Result<(), Error> {
+        let s = match self.streams_by_id.get_mut(&stream_id) {
+            None => bail!(Internal, msg("no stream {stream_id}")),
+            Some(s) => s,
+        };
+        let dir_id = match s.sample_file_dir_id {
+            None => bail!(Internal, msg("stream {stream_id} has no directory!")),
+            Some(d) => d,
+        };
+
+        let composite_ids = CompositeId::new(stream_id, recording_ids.start)..CompositeId::new(stream_id, recording_ids.end);
+
+        {
+            let tx = self.conn.transaction()?;
+            match raw::delete_recordings(&tx, dir_id, composite_ids)? {
+                0 => bail!(Internal, msg("zero recordings deleted from {stream_id}/{recording_ids:?}")),
+                _ => {},
+            };
+            tx.commit()?;
+        };
+        self.init_sample_file_dirs_garbage_needs_unlink()?;
```

Comment:
I think I would do something like `let new_garbage_needs_unlink = raw::list_garbage(&tx, dir_id)?;` within the transaction above, then assign to dirs post-commit if the transaction is successful. Then we don't have any possibility that the write succeeds but the re-init doesn't happen.

I had to refresh my memory on how this works as it's been quite a while. It looks like the actual unlinking should happen automatically shortly after next flush. We might call out in the API docs that it's asynchronous; folks might otherwise expect the disk space to be available by the time the HTTP request finishes.